src/System.Memory/src/System/Buffers/ReadOnlySequence.cs

```diff
@@ -330,15 +330,29 @@ private ReadOnlySequence<T> SliceImpl(in SequencePosition begin, in SequencePosi
             );
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private SequenceType GetSequenceType()
+        {
+            // We take high order bits of two indexes and move them
+            // to a first and second position to convert to SequenceType
+            return _sequenceStart.GetObject() == null
+                ? SequenceType.Empty
+                : (SequenceType)(-(2 * (_sequenceStart.GetInteger() >> 31) + (_sequenceEnd.GetInteger() >> 31)));
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int GetIndex(in SequencePosition position) => position.GetInteger() & ReadOnlySequence.IndexBitMask;
+
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private void GetTypeAndIndices(int start, int end, out SequenceType sequenceType, out int startIndex, out int endIndex)
         {
             startIndex = start & ReadOnlySequence.IndexBitMask;
             endIndex = end & ReadOnlySequence.IndexBitMask;
-            // We take high order bits of two indexes index and move them
-            // to a first and second position to convert to BufferType
-            // Masking with 2 is required to only keep the second bit of Start.GetInteger()
-            sequenceType = Start.GetObject() == null ? SequenceType.Empty : (SequenceType)((((uint)Start.GetInteger() >> 30) & 2) | (uint)End.GetInteger() >> 31);
+            // We take high order bits of two indexes and move them
+            // to a first and second position to convert to SequenceType
+            sequenceType = _sequenceStart.GetObject() == null
```

Comment:
Call the GetSequenceType instead, to avoid code duplication
`sequenceType = GetSequenceType();`